File: create_public_bucket.py
from __future__ import annotations

import json
import os

from minio import Minio
from minio.error import S3Error

from dotenv import load_dotenv
load_dotenv()

# Minio server and credentials
minio_endpoint = os.getenv("AWS_S3_ENDPOINT")
access_key = os.getenv("AWS_ACCESS_KEY_ID")
secret_key = os.getenv("AWS_SECRET_ACCESS_KEY")
bucket_name = os.getenv("AWS_STORAGE_BUCKET_NAME")

print(
    "******",
    minio_endpoint,
    access_key,
    secret_key,
)

# Create a Minio client
minio_client = Minio(
    minio_endpoint,
    access_key=access_key,
    secret_key=secret_key,
    secure=False,
)


# Create the bucket if it doesn't exist
found = minio_client.bucket_exists(bucket_name)
if not found:
    minio_client.make_bucket(bucket_name)
    print(f"Bucket ${bucket_name} created.")
else:
    print(f"Bucket ${bucket_name} already exists.")

# Set the bucket policy to public

policy = {
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {"AWS": "*"},
            "Action": "s3:GetObject",
            "Resource": f"arn:aws:s3:::{bucket_name}/*",
        },
    ],
}

try:
    minio_client.set_bucket_policy(bucket_name, json.dumps(policy))
    print(f'Bucket policy for "{bucket_name}" set to public')
except S3Error as e:
    print(f"Error setting bucket policy: {str(e)}")

File: create_public_bucket.js
const Minio = require("minio");
const dotenv = require("dotenv");

dotenv.config();

// Minio server and credentials
const minioEndpoint = process.env.AWS_S3_ENDPOINT;
const accessKey = process.env.AWS_ACCESS_KEY_ID;
const secretKey = process.env.AWS_SECRET_ACCESS_KEY;
const bucketName = process.env.AWS_STORAGE_BUCKET_NAME;

console.log("******", minioEndpoint, accessKey, secretKey);
const [endPoint, port] = minioEndpoint.split(":");

// Create a Minio client
const minioClient = new Minio.Client({
  port: +port,
  endPoint,
  useSSL: false,
  accessKey: accessKey,
  secretKey: secretKey,
});

// Create the bucket if it doesn't exist
minioClient.bucketExists(bucketName, function (err, found) {
  if (err) {
    return console.log(err);
  }

  if (!found) {
    minioClient.makeBucket(bucketName, "", function (err) {
      if (err) {
        return console.log(`Error creating bucket ${bucketName}: ${err}`);
      }
      console.log(`Bucket ${bucketName} created.`);
    });
  } else {
    console.log(`Bucket ${bucketName} already exists.`);
  }
});

// Set the bucket policy to public
const policy = {
  Version: "2012-10-17",
  Statement: [
    {
      Effect: "Allow",
      Principal: { AWS: "*" },
      Action: "s3:GetObject",
      Resource: `arn:aws:s3:::${bucketName}/*`,
    },
  ],
};

minioClient.setBucketPolicy(bucketName, JSON.stringify(policy), function (err) {
  if (err) {
    return console.log(`Error setting bucket policy: ${err}`);
  }
  console.log(`Bucket policy for "${bucketName}" set to public`);
});
